worker/src/d1/mod.rs

```diff
@@ -0,0 +1,228 @@
+        } else {
+            Err(format!(
+                "Binding cannot be cast to the type {} from {}",
+                Self::TYPE_NAME,
+                obj.constructor().name()
+            )
+            .into())
+        }
+    }
+}
+
+impl JsCast for Database {
+    fn instanceof(val: &JsValue) -> bool {
+        val.is_instance_of::<D1DatabaseSys>()
+    }
+
+    fn unchecked_from_js(val: JsValue) -> Self {
+        Self(val.into())
+    }
+
+    fn unchecked_from_js_ref(val: &JsValue) -> &Self {
+        unsafe { &*(val as *const JsValue as *const Self) }
+    }
+}
+
+impl From<Database> for JsValue {
+    fn from(database: Database) -> Self {
+        JsValue::from(database.0)
+    }
+}
+
+impl AsRef<JsValue> for Database {
+    fn as_ref(&self) -> &JsValue {
+        &self.0
+    }
+}
+
+impl From<D1DatabaseSys> for Database {
+    fn from(inner: D1DatabaseSys) -> Self {
+        Self(inner)
+    }
+}
+
+// A D1 prepared query statement.
+pub struct PreparedStatement(D1PreparedStatementSys);
+
+impl PreparedStatement {
+    /// Bind one or more parameters to the statement.
+    /// Consumes the old statement and returns a new statement with the bound parameters.
+    ///
+    /// D1 follows the SQLite convention for prepared statements parameter binding.
+    ///
+    /// # Considerations
+    ///
+    /// Supports Ordered (?NNNN) and Anonymous (?) parameters - named parameters are currently not supported.
+    ///
+    pub fn bind(self, values: &[JsValue]) -> Result<Self> {
+        let array: Array = values.iter().collect::<Array>();
+
+        match self.0.bind(array) {
+            Ok(stmt) => Ok(PreparedStatement(stmt)),
+            Err(err) => Err(Error::from(err)),
+        }
+    }
+
+    /// Return the first row of results.
+    ///
+    /// If `col_name` is `Some`, returns that single value, otherwise returns the entire object.
+    ///
+    /// If the query returns no rows, then this will return `None`.
+    ///
+    /// If the query returns rows, but column does not exist, then this will return an `Err`.
+    pub async fn first<T>(&self, col_name: Option<&str>) -> Result<Option<T>>
+    where
+        T: for<'a> Deserialize<'a>,
+    {
+        let js_value = JsFuture::from(self.0.first(col_name)).await?;
+        let value = serde_wasm_bindgen::from_value(js_value)?;
+        Ok(value)
+    }
+
+    /// Executes a query against the database but only return metadata.
+    pub async fn run(&self) -> Result<D1Result> {
+        let result = JsFuture::from(self.0.run()).await?;
+        Ok(D1Result(result.into()))
+    }
+
+    /// Executes a query against the database and returns all rows and metadata.
+    pub async fn all(&self) -> Result<D1Result> {
+        let result = JsFuture::from(self.0.all()).await?;
+        Ok(D1Result(result.into()))
+    }
+
+    /// Executes a query against the database and returns a `Vec` of rows instead of objects.
+    pub async fn raw<T>(&self) -> Result<Vec<Vec<T>>>
```

Comment:
I don't think it should take a generic type T because consumers will unlikely write a select statement where all the types are the same. In the people case, select * from my table will cause a type error between the String, Integer or other types defined in the table's column.  Perhaps, it should be a String rather than T?

Reply:
This is true - I'll take a look. I think this was supposed to be a `Result<Vec<T>>`, but not sure.